src/lib/mina_ledger/ledger.ml

```diff
@@ -201,6 +277,47 @@ module Ledger_inner = struct
     let _base, mask = create_ephemeral_with_base ~depth () in
     mask
 
+  type converting_config =
+    { primary_directory_name : string option
+    ; converting_directory_name : string option
+    }
+
+  let default_converting_directory_name primary_directory_name =
+    primary_directory_name ^ "_converting"
+
+  let converting_directory_name ~cfg ~primary_directory =
+    Option.first_some cfg.converting_directory_name
+    @@ Option.map primary_directory ~f:default_converting_directory_name
+
+  let empty_converting_config : converting_config =
+    { primary_directory_name = None; converting_directory_name = None }
+
+  let create_converting ?(cfg = empty_converting_config) ~depth () =
+    let open Or_error.Let_syntax in
+    let db1 = Db.create ?directory_name:cfg.primary_directory_name ~depth () in
+    let db2 =
+      let directory_name =
+        converting_directory_name ~cfg ~primary_directory:(Db.get_directory db1)
+      in
+      Unstable_db.create ?directory_name ~depth ()
+    in
+    let%bind converting_ledger =
+      if Unstable_db.num_accounts db2 = 0 then
+        Or_error.return @@ Converting_ledger.create_with_migration db1 db2
+      else if not (Db.num_accounts db1 = Unstable_db.num_accounts db2) then (
+        (* TODO: ensure 2 dbs have exact same accounts *)
+        Db.close db1 ;
+        Unstable_db.close db2 ;
+        Or_error.error_string
```

Comment:
I think if a desync is detected then we might want to clean up the secondary database and re-migrate, and just issue a warning/error log message. I'm not sure if that cleanup should happen here or if it should be the responsibility of the caller of this method, though. I'd have to check to see how the existing databases are cleaned up.

(The issue I see is similar to the crash back-off loop with the empty PID file - if the invalid resource isn't cleaned up automatically somehow then the working directory of the daemon will become unusable).